internal/engine/compiler/impl_amd64.go

```diff
@@ -3598,21 +3598,42 @@ func (c *amd64Compiler) compileLoadDataInstanceAddress(dataIndex uint32, dst asm
 	)
 }
 
-// compileMemoryCopy implements compiler.compileMemoryCopy for the amd64 architecture.
-func (c *amd64Compiler) compileMemoryCopy() error {
-	return c.compileCopyImpl(false, 0, 0)
-}
+func (c *amd64Compiler) compileCopyLoopImpl(sourceOffset, destinationOffset, copySize asm.Register, backwards bool, bwOffset uint8) {
+	// skip if nothing to copy
+	c.assembler.CompileRegisterToConst(amd64.CMPQ, copySize, 0)
+	emptyEightGroupsJump := c.assembler.CompileJump(amd64.JEQ)
 
-// compileCopyImpl implements compileTableCopy and compileMemoryCopy.
-//
-// TODO: the compiled code in this function should be reused and compile at once as
-// the code is independent of any module.
-func (c *amd64Compiler) compileCopyImpl(isTable bool, srcTableIndex, dstTableIndex uint32) error {
-	outOfBoundsErrorStatus := nativeCallStatusCodeMemoryOutOfBounds
-	if isTable {
-		outOfBoundsErrorStatus = nativeCallStatusCodeInvalidTableAccess
+	// prepare registers for REP MOVSQ: copy from rsi to rdi, rcx times
+	c.assembler.CompileRegisterToRegister(amd64.XCHGQ, destinationOffset, amd64.RegDI)
+	c.assembler.CompileRegisterToRegister(amd64.XCHGQ, sourceOffset, amd64.RegSI)
+	c.assembler.CompileRegisterToRegister(amd64.XCHGQ, copySize, amd64.RegCX)
```

Comment:
It was silly for me to assume this will work. If we have DI/SI/CX among our used registers this might cause a double `XCHG` where we get wrong values.
We might either push them on the stack (slow) or invent something clever. I bet any case can be resolved with 3 XCHGS or less 🤔